yapapi/services.py

```diff
@@ -191,26 +204,159 @@ def _respond_nowait(self, message: Optional[Any], response_to: Optional[ServiceS
+        it's expected to be a generator that yields `WorkItems` (generated using the service's
+        instance of the work context) that are then dispatched to the activity by the engine.
+
+        Results of those batches can then be retrieved by awaiting the values captured from yield
+        statements.
+
+        A clean exit from a handler function triggers the engine to transition the state of the
+        instance to the next stage in service's lifecycle - in this case, to `stopping`.
+
+        Any unhandled exception will cause the instance to be terminated.
+
+        ### Example
+
+        ```
+        async def run(self):
+            while True:
+                self._ctx.run("/golem/run/report", "--stats")  # index 0
+                future_results = yield self._ctx.commit()
+                results = await future_results
+                stats = results[0].stdout.strip()  # retrieve from index 0
+                print(f"stats: {stats}")
+        ```
+
+        ### Default implementation
+
+        Because the nature of the operations required during the "running" state depends directly
+        on the specifics of a given Service and because it's entirely plausible for a service
+        not to require any direct interaction with the exe-unit (runtime) from the requestor's end
+        after the service has been started, the default is to just wait indefinitely without
+        producing any batches.
+        """
 
         await asyncio.Future()
-        yield
+        yield  # type: ignore # unreachable because of the indefinite wait above
 
-    async def shutdown(self):
-        """Implement the `stopping` state of the service."""
+    async def shutdown(self) -> AsyncGenerator[WorkItem, Awaitable[List[events.CommandEvent]]]:
+        """Implement the handler for the `stopping` state of the service.
+
+        To be overridden by the author of a specific Service class.
+
+        Should contain any operations that the requestor needs to ensure the instance is correctly
+        shut-down - e.g. that its final state is retrieved.
+
+        As a handler implementing the [work generator pattern](https://handbook.golem.network/requestor-tutorials/golem-application-fundamentals/hl-api-work-generator-pattern),
+        it's expected to be a generator that yields `WorkItems` (generated using the service's
+        instance of the work context) that are then dispatched to the activity by the engine.
+
+        Results of those batches can then be retrieved by awaiting the values captured from yield
+        statements.
+
+        Finishing the execution of this handler will trigger termination of this instance.
+
+        This handler will only be called if the activity running the service is still available.
+        If the activity has already been deemed terminated or if the connection with the provider
```

Comment:
```suggestion
        If the activity has already been terminated or if the connection with the provider
```

Reply:
I'd leave "deemed" to convey the notion that it can happen outside the requestor's control